prisma-fmt/src/references.rs

```diff
@@ -0,0 +1,237 @@
+        ) => find_where_used_as_field_name(&ctx, arg_val.as_str(), model_id, ctx.initiating_file_id)
+            .into_iter()
+            .collect(),
+
+        _ => vec![],
+    };
+
+    identifiers
+        .iter()
+        .filter_map(|ident| ident_to_location(ident, &ctx))
+        .collect()
+}
+
+fn find_where_used_as_field_name<'ast>(
+    ctx: &'ast ReferencesContext<'_>,
+    name: &str,
+    model_id: ModelId,
+    file_id: FileId,
+) -> Option<&'ast Identifier> {
+    let model = ctx.db.walk((file_id, model_id));
+
+    match model.scalar_fields().find(|field| field.name() == name) {
+        Some(field) => Some(field.ast_field().identifier()),
+        None => None,
+    }
+}
+
+fn find_where_used_for_native_type<'ast>(
+    ctx: &ReferencesContext<'ast>,
+    name: &'ast str,
+) -> impl Iterator<Item = &'ast Identifier> {
+    fn find_native_type_locations<'ast>(
+        name: &'ast str,
+        fields: impl Iterator<Item = (FieldId, &'ast Field)> + 'ast,
+    ) -> Box<dyn Iterator<Item = &'ast Identifier> + 'ast> {
+        Box::new(fields.filter_map(move |field| {
+            field
+                .1
+                .attributes
+                .iter()
+                .find(|attr| extract_ds_from_native_type(attr.name()) == name)
+                .map(|attr| attr.identifier())
+        }))
+    }
+
+    ctx.db.walk_tops().flat_map(move |top| match top.ast_top() {
+        Top::CompositeType(composite_type) => find_native_type_locations(name, composite_type.iter_fields()),
+        Top::Model(model) => find_native_type_locations(name, model.iter_fields()),
+
+        Top::Enum(_) | Top::Source(_) | Top::Generator(_) => Box::new(empty_identifiers()),
+    })
+}
+
+fn find_where_used_as_field_type<'ast>(
+    ctx: &'ast ReferencesContext<'_>,
+    name: &'ast str,
+) -> impl Iterator<Item = &'ast Identifier> {
+    fn get_relevent_identifiers<'a>(
+        fields: impl Iterator<Item = (FieldId, &'a Field)>,
+        name: &str,
+    ) -> Vec<&'a Identifier> {
+        fields
+            .filter_map(|(_id, field)| match &field.field_type {
+                FieldType::Supported(id) if id.name == name => Some(id),
+                _ => None,
+            })
+            .collect()
+    }
+
+    ctx.db.walk_tops().flat_map(|top| match top.ast_top() {
+        Top::Model(model) => get_relevent_identifiers(model.iter_fields(), name),
+        Top::CompositeType(composite_type) => get_relevent_identifiers(composite_type.iter_fields(), name),
+        // * Cannot contain field types
+        Top::Enum(_) | Top::Source(_) | Top::Generator(_) => vec![],
+    })
+}
+
+fn find_where_used_as_top_name<'ast>(ctx: &'ast ReferencesContext<'_>, name: &'ast str) -> Option<&'ast Identifier> {
+    ctx.db.find_top(name).map(|top| top.ast_top().identifier())
+}
+
+fn extract_ds_from_native_type(attr_name: &str) -> &str {
+    attr_name.split('.').collect::<Vec<&str>>()[0]
+}
```

Comment:
This is still collecting a vec instead of consuming the iterator

Reply:
Ah!